src/collection.ts

```diff
@@ -1,104 +1,103 @@
 import GitDB from './gitdb';
 import { v4 as uuidv4 } from 'uuid';
-import fse from 'fs-extra';
-interface Record {
+import { FileStrategy } from './FileStrategy';
+import { MemoryStrategy } from './MemoryStrategy';
+export interface DBRecord {
   id: string;
 }
 
-type Filter<T> = (document: T) => boolean;
-
-type SetCallback<T> = (record: T) => T;
-type UpdateOrCallback<T, K extends T> = T | SetCallback<K>;
-export enum WriteToDisk {
-  DO_NOT_WRITE_TO_DISK = 0,
-  WRITE_TO_DISK = 1,
+export type Filter<T> = (document: T) => boolean;
+export type SetCallback<T> = (record: T) => T;
+enum gitStagingAreaStatus {
+  add,
+  remove,
 }
-
 /**
  * TODO add immutablejs to prevent array and object mutation
  */
-export default class Collection<T extends Record> {
+export default class Collection<T extends DBRecord> {
   private db: GitDB;
-  private name: string;
-  private data: any[];
-  get length(): number {
-    return this.data.length;
-  }
+  fileStrategy: FileStrategy<T>;
+  memoryStrategy: MemoryStrategy<T>;
 
-  constructor(gitDb: GitDB, name: string) {
+  constructor(
+    gitDb: GitDB,
+    fileStrategy: FileStrategy<T>,
+    memoryStrategy: MemoryStrategy<T>,
+  ) {
     this.db = gitDb;
-    this.name = name;
-    this.data = [];
+    this.fileStrategy = fileStrategy;
+    this.memoryStrategy = memoryStrategy;
   }
 
-  public async getData(callback: Filter<T>): T[] {
-    const collection = this.db.config.cache
-      ? this
-      : await this.db.readCollection(this.name, true);
+  public async getAll(): Promise<T[]> {
+    const documents = this.db.config.cache
+      ? this.memoryStrategy.getAll()
+      : this.fileStrategy.getAll();
+    return documents;
+  }
 
-    if (typeof callback === 'function') {
-      return collection.data.filter(callback);
-    } else {
-      return collection.data;
-    }
+  public async getData(callback: Filter<T>): Promise<T[]> {
+    const filteredDocuments = this.db.config.cache
+      ? this.memoryStrategy.getData(callback)
+      : this.fileStrategy.getData(callback);
+
+    return filteredDocuments;
   }
 
-  public async insert(
-    documentData: T,
-    writeToDisk: WriteToDisk = WriteToDisk.WRITE_TO_DISK,
-  ): Promise<T> {
+  public async insert(documentData: T): Promise<T> {
     const newDocument = { id: uuidv4(), ...documentData };
+    const filePath = await this.fileStrategy.insert(newDocument);
+
     if (this.db.config.cache) {
```

Comment:
checking if `this.memoryStrategy` is present might be better indicator whether something should be cached. `this.db.config.cache` might not be required for collection at all. 

Maybe we should control whether memory strategy should be passed in GitDB based on `db.config.cache`